lib/Parse/Lexer.cpp

```diff
@@ -1206,6 +1206,38 @@ static bool maybeConsumeNewlineEscape(const char *&CurPtr, ssize_t Offset) {
   }
 }
 
+/// diagnoseZeroWidth - check for and error zero-width characters in delimiters
+static bool diagnoseZeroWidth(const char *&CurPtr, DiagnosticEngine *Diags) {
+  const unsigned char *TmpPtr = (const unsigned char *)CurPtr;
+  // zero-width set assumed to be: U+200B, U+200C, U+200D, U+2060, U+FEFF
```

Comment:
Thanks for your comments. I’ll work through them tomorrow. I looked at supporting ignoring invisible characters rather than reporting an error and it would be _much_ harder to implement as the relationship between the token span and the actual literal will become +/- a variable length rather than just (1 or 3) + DelimiterLength. Besides I don’t think it's a good idea - more prone to bugs though I see where you are coming from with other whitespace.

I figured there would be a difficult to anticipate set of zero-width chars to exclude which is going to make this difficult to make watertight. I think this is an implementation detail of diagnoseZeroWidth() which I’d like to leave in this PR in its incomplete but architected form while we’re working in the neighbourhood as a start for more comprehensive treatment by someone else later. Not sure how to use ICU but doesn’t that requite 16 bit chars? So I’d like to not revert the last commit if that’s ok before layering the next commit on top of it.

Reply:
Personally, I feel _very strongly_ against leaving this commit in. It's not addressing the issue in any systematic way but hides the easiest-to-demonstrate manifestation of it.

Either we believe that this is not a major issue or we believe that it is. In the former case, we can fix this for all multi-character string delimiters in some systematic way at a later time. In the latter case, we _must_ fix this for all multi-character string delimiters in some systematic way as soon as possible. But in either case, I can see no justification for a partial implementation now, which has the problem of hiding the issue but not solving it.